### PR TITLE
Make breakpoints use the token feature

### LINKED ISSUE #32954: 响应式栅格可否增加1920的尺寸？
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
目前的响应式栅格xxl 是≥ 1600px，但是现在好多1920的屏幕有好多，2560的也有很多，从1920到2560直接是否也加一下，例如加一个xxxl的尺寸是>=2000

### What does the proposed API look like?
xxxl: >=2000px

### PR DESCRIPTION
**Pull request by azro352**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [X] New feature

### 🔗 Related issue link

close https://github.com/ant-design/ant-design/issues/32954
close https://github.com/ant-design/ant-design/issues/34197
close https://github.com/ant-design/ant-design/issues/19694
close https://github.com/ant-design/ant-design/issues/22359

### 💡 Background and solution

Screens are getting bigger, we need a new size to handle them. For example with 4k screens, you can differenciate an half-screen of a full-screen (so can't adapt the width f you app only for full-screen mode)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Breakpoints can now follow theme token config. |
| 🇨🇳 Chinese | Breakpoints 现在会消费主题 token 配置。 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated
- [X] Demo is updated
- [X] TypeScript definition is updated
- [X] Changelog is provided or not needed

The decision has been made to use the new token system for the breakpoints

I have decided to remove `ResponsiveObserve.unsubscribe should be called when unmounted` test because now we use a hook, that isn't the same instance of `ResponsiveObserver` that is shared between Row and the test, so we can't check the call on subscribe at this moment

Thanks

### CODE REVIEW

**Review by yoyo837 — APPROVED** [review]

**Comment by yoyo837** [comment]
Maybe `feature` branch is better.

**Comment by yoyo837** [comment]
CI failed.

**Comment by yoyo837** [comment]
`feature` branch please!

**Comment by azro352** [comment]
> `feature` branch please!

You said "maybe" so I wasn't sure If I needed to change or not , sorry :)

**Comment by yoyo837** [comment]
> > `feature` branch please!
> 
> You said "maybe" so I wasn't sure If I needed to change or not , sorry :)

My bad. :)

File: components/list/index.zh-CN.md [hunk 1]
```diff
@@ -70,6 +70,7 @@ cover: https://gw.alipayobjects.com/zos/alicdn/5FrZKStG_/List.svg
 | lg     | `≥992px` 展示的列数  | number | -      |      |
 | xl     | `≥1200px` 展示的列数 | number | -      |      |
 | xxl    | `≥1600px` 展示的列数 | number | -      |      |
+| xxxl     | `≥2000px` column of grid   | number | -       | 5.0.x |
```
Comment (huangkairan) [inline components/list/index.zh-CN.md:73]: | xxxl     | `≥2000px` 展示的列数   | number | -       | 5.0.x |

File: components/grid/demo/responsive.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ## zh-CN
 
-参照 Bootstrap 的 [响应式设计](http://getbootstrap.com/css/#grid-media-queries)，预设六个响应尺寸：`xs` `sm` `md` `lg` `xl` `xxl`。
+参照 Bootstrap 的 [响应式设计](http://getbootstrap.com/css/#grid-media-queries)，预设六个响应尺寸：`xs` `sm` `md` `lg` `xl` `xxl` `xxxl`。
```
Comment (li-jia-nan) [inline components/grid/demo/responsive.md:3]: `六个响应尺寸` should is `七个响应尺寸`

File: components/grid/demo/responsive.md [hunk 3]
```diff
@@ -1,7 +1,7 @@
...
 
 ## en-US
 
-Referring to the Bootstrap [responsive design](http://getbootstrap.com/css/#grid-media-queries), here preset six dimensions: `xs` `sm` `md` `lg` `xl` `xxl`.
+Referring to the Bootstrap [responsive design](http://getbootstrap.com/css/#grid-media-queries), here preset six dimensions: `xs` `sm` `md` `lg` `xl` `xxl` `xxxl`.
```
Comment (li-jia-nan) [inline components/grid/demo/responsive.md:7]: `six dimensions` should is `seven dimensions`

File: components/grid/demo/responsive.md (hunk 2 above)
Comment (li-jia-nan) [inline components/grid/demo/responsive.md:3]:
```suggestion
参照 Bootstrap 的 [响应式设计](http://getbootstrap.com/css/#grid-media-queries)，预设七个响应尺寸：`xs` `sm` `md` `lg` `xl` `xxl` `xxxl`。
```

File: components/grid/demo/responsive.md (hunk 3 above)
Comment (li-jia-nan) [inline components/grid/demo/responsive.md:7]:
```suggestion
Referring to the Bootstrap [responsive design](http://getbootstrap.com/css/#grid-media-queries), here preset seven dimensions: `xs` `sm` `md` `lg` `xl` `xxl` `xxxl`.
```

File: components/list/index.zh-CN.md (hunk 1 above)
Comment (afc163) [reply components/list/index.zh-CN.md:73]: Why 2000px?

Comment (azro352) [reply components/list/index.zh-CN.md]: @afc163 I took the value suggested by the [issue-xxxl](https://github.com/ant-design/ant-design/issues/32954) and also used by the https://github.com/vueComponent/ant-design-vue/pull/4953 version

File: components/list/index.zh-CN.md [hunk 4]
```diff
@@ -70,7 +70,7 @@ cover: https://gw.alipayobjects.com/zos/alicdn/5FrZKStG_/List.svg
 | lg     | `≥992px` 展示的列数  | number | -      |      |
 | xl     | `≥1200px` 展示的列数 | number | -      |      |
 | xxl    | `≥1600px` 展示的列数 | number | -      |      |
-| xxxl     | `≥2000px` 展示的列数 of grid   | number | -       | 5.0.x |
+| xxxl   | `≥2000px` 展示的列数 | number | -       | 5.0.x |
```
Comment (yoyo837) [inline components/list/index.zh-CN.md:73]:
```suggestion
| xxxl   | `≥2000px` 展示的列数 | number | -       | 5.1.0 |
```

File: components/grid/index.zh-CN.md [hunk 5]
```diff
@@ -58,27 +58,28 @@ Ant Design 的布局组件若不能满足你的需求，你也可以直接使用
...
+| pull   | 栅格向左移动格数                              | number | 0 |  |
+| push   | 栅格向右移动格数                              | number | 0 |  |
+| span   | 栅格占位格数，为 0 时相当于 `display: none`       | number | - |  |
+| xs     | `屏幕 < 576px` 响应式栅格，可为栅格数或一个包含其他属性的对象  | number \| object | - |  |
+| sm     | `屏幕 ≥ 576px` 响应式栅格，可为栅格数或一个包含其他属性的对象  | number \| object | - |  |
+| md     | `屏幕 ≥ 768px` 响应式栅格，可为栅格数或一个包含其他属性的对象  | number \| object | - |  |
+| lg     | `屏幕 ≥ 992px` 响应式栅格，可为栅格数或一个包含其他属性的对象  | number \| object | - |  |
+| xl     | `屏幕 ≥ 1200px` 响应式栅格，可为栅格数或一个包含其他属性的对象 | number \| object | - |  |
+| xxl    | `屏幕 ≥ 1600px` 响应式栅格，可为栅格数或一个包含其他属性的对象 | number \| object | - |  |
+| xxxl   | `屏幕 ≥ 2000px` 响应式栅格，可为栅格数或一个包含其他属性的对象    | number \| object | - | 5.0.x |
```
Comment (yoyo837) [inline components/grid/index.zh-CN.md:82]:
```suggestion
| xxxl   | `屏幕 ≥ 2000px` 响应式栅格，可为栅格数或一个包含其他属性的对象    | number \| object | - | 5.1.0 |
```

File: components/list/index.en-US.md [hunk 6]
```diff
@@ -69,6 +69,7 @@ More about pagination, please check [`Pagination`](/components/pagination/).
 | lg       | `≥992px` column of grid  | number | -       |         |
 | xl       | `≥1200px` column of grid | number | -       |         |
 | xxl      | `≥1600px` column of grid | number | -       |         |
+| xxxl     | `≥2000px` column of grid   | number | -       | 5.0.x |
```
Comment (yoyo837) [inline components/list/index.en-US.md:72]:
```suggestion
| xxxl     | `≥2000px` column of grid   | number | -       | 5.1.0 |
```

File: components/_util/responsiveObserve.ts [hunk 7]
```diff
@@ -12,6 +12,7 @@ export const responsiveMap: BreakpointMap = {
   lg: '(min-width: 992px)',
   xl: '(min-width: 1200px)',
   xxl: '(min-width: 1600px)',
+  xxxl: '(min-width: 2000px)',
```
Comment (zombieJ) [inline components/_util/responsiveObserve.ts:15]: Too many break point.
This seems not a good design. Which should be modified to read from [token](https://github.com/ant-design/ant-design/blob/master/components/theme/util/alias.ts#L27) and set by ConfigProvider with `theme` instead.

Comment (azro352) [reply components/_util/responsiveObserve.ts:15]: If that is the decision you make guys, I don't know if I'll be able to make the changes myself, I don't really understand what you're talking about, and still less how to implement it :/

**Review by yoyo837 — DISMISSED** [review]
So sorry, The owner think that this should be supported by adding some style token by ConfigProvider with theme. https://github.com/ant-design/ant-design/pull/39105#discussion_r1035997601
https://ant.design/docs/react/customize-theme#customize-design-token

File: components/_util/responsiveObserve.ts (hunk 7 above)
Comment (azro352) [reply components/_util/responsiveObserve.ts:15]: @yoyo837 
I'll try something, do I need to make something like https://github.com/ant-design/ant-design/blob/master/components/grid/style/index.tsx#L139 and line 150 ?

Comment (zombieJ) [reply components/_util/responsiveObserve.ts:15]: As above mentioned. It should be able to reconfig break point with ConfigProvider:
```tsx
<ConfigProvider
  theme={{
    token: {
      screenXS: xxx, // Some new value
      ...
    },
  }}
>
</ConfigProvider>
```

For this, current break point not fully control by theme token. It still exist in some part outside: https://github.com/ant-design/ant-design/blob/77ea4038d169d033dcd4d29da1df9ab21b23ebb9/components/_util/responsiveObserve.ts#L8

What should do is modify up code to support consume ConfigProvider token.

Comment (azro352) [reply components/_util/responsiveObserve.ts:15]: After reading more code, it seems that `genComponentStyleHook` is not what I need, I think that is something like the following is needed, but I don't where to get the token from 
```js
// 'fromToken' doesn't exist, it is just to show what I think I need 

const useResponsiveStyle = fromToken((token) => {
  return {
    'xs': `(max-width: ${token.screenXS}px)`,
    'sm': `(max-width: ${token.screenSM}px)`,
    'md': `(max-width: ${token.screenMD}px)`,
    'lg': `(max-width: ${token.screenLG}px)`,
    'xl': `(max-width: ${token.screenXL}px)`,
    'xxl': `(max-width: ${token.screenXXL}px)`,
    'xxxl': `(max-width: ${token.screenXXXL}px)`,
  };
});

export const responsiveMap: BreakpointMap = useResponsiveStyle();
```

I'm ok to continue on this PR to make this work, but I need some guidance please :)

Comment (azro352) [reply components/_util/responsiveObserve.ts:15]: > As above mentioned. It should be able to reconfig break point with ConfigProvider:
> ...
> What should do is modify up code to support consume ConfigProvider token.

@zombieJ  That I got it now. But I don't find what method to use for, can you point me some place where that thing is done (or maybe some documentation about how to implement it) ?

**Comment by azro352** [comment]
@zombieJ @yoyo837 

Is that something like this in `compoments/_util/responsiveObserve.ts` file ? 

```js
import { useToken } from '../../theme/internal';

const [theme, token, hashId] = useToken();

export const responsiveMap: BreakpointMap = {
  xs: `(max-width: ${token.screenXS}px)`,
  sm: `(max-width: ${token.screenSM}px)`,
  md: `(max-width: ${token.screenMD}px)`,
  lg: `(max-width: ${token.screenLG}px)`,
  xl: `(max-width: ${token.screenXL}px)`,
  xxl: `(max-width: ${token.screenXXL}px)`,
  xxxl: `(max-width: ${token.screenXXXL}px)`,
};
```


Or if useToken needs to be called in a metho

````js
export const responsiveMap = (): BreakpointMap => {
  const [theme, token, hashId] = useToken();
  return {
    xs: `(max-width: ${token.screenXS}px)`,
    sm: `(max-width: ${token.screenSM}px)`,
    md: `(max-width: ${token.screenMD}px)`,
    lg: `(max-width: ${token.screenLG}px)`,
    xl: `(max-width: ${token.screenXL}px)`,
    xxl: `(max-width: ${token.screenXXL}px)`,
    xxxl: `(max-width: ${token.screenXXXL}px)`,
  };
}();
```

File: components/_util/responsiveObserve.ts (hunk 7 above)
Comment (azro352) [reply components/_util/responsiveObserve.ts]: @zombieJ @yoyo837 

Is that something like this in `compoments/_util/responsiveObserve.ts` file ? 

```js
import { useToken } from '../../theme/internal';

const getResponsiveMap = (): BreakpointMap => {
  const [, token, ] = useToken();
  return {
    xs: `(max-width: ${token.screenXS}px)`,
    sm: `(max-width: ${token.screenSM}px)`,
    md: `(max-width: ${token.screenMD}px)`,
    lg: `(max-width: ${token.screenLG}px)`,
    xl: `(max-width: ${token.screenXL}px)`,
    xxl: `(max-width: ${token.screenXXL}px)`,
    xxxl: `(max-width: ${token.screenXXXL}px)`,
  } as BreakpointMap;
};

export const responsiveMap: BreakpointMap = getResponsiveMap();
```

Comment (zombieJ) [reply components/_util/responsiveObserve.ts]: origin `compoments/_util/responsiveObserve.ts` export default with `responsiveObserve` object which do not have the ability to consume token context. It should change to hooks version:

```diff
  // compoments/_util/responsiveObserve.ts

- export default responsiveObserve;
+ export default function useResponsiveObserve() {
+   const [, token] = useToken();
+   // ...
+ }
```

And then adjust the code which use `responsiveObserve` to hooks usage.

Comment (azro352) [reply components/_util/responsiveObserve.ts]: @zombieJ That would make a lots of changes, in many files

Don't my actual idea would fit, just make the responsiveMap a hook ? 

```js
const useResponsiveMap = (): BreakpointMap => {
  const [, token] = useToken();
  return {
    xs: `(max-width: ${token.screenXS}px)`,
    sm: `(max-width: ${token.screenSM}px)`,
    md: `(max-width: ${token.screenMD}px)`,
    lg: `(max-width: ${token.screenLG}px)`,
    xl: `(max-width: ${token.screenXL}px)`,
    xxl: `(max-width: ${token.screenXXL}px)`,
    xxxl: `(max-width: ${token.screenXXXL}px)`,
  } as BreakpointMap;
};
```

That only causes 2 changes in the same file 
```js
  unregister() {
    const responsiveMap: BreakpointMap = useResponsiveMap();
    Object.keys(responsiveMap).forEach((screen: Breakpoint) => {
...
  register() {
    const responsiveMap: BreakpointMap = useResponsiveMap();
    Object.keys(responsiveMap).forEach((screen: Breakpoint) => {
```

Comment (azro352) [reply components/_util/responsiveObserve.ts]: @zombieJ I've done something , I let you see if that corresponds to your expectation

Basically, I've done this :

```diff
-   const token = ResponsiveObserve.subscribe((screen) => {
+  const responsiveObserve = useResponsiveObserve();
+  const token = responsiveObserve.subscribe((screen) => {
```

**Comment by azro352** [comment]
Since I merged `feature` into my branch, to fix the conflicts, it seems I got error not related to my code , but I'm unsure. I try running the tests in local but they take so much time

**Comment by afc163** [comment]
Adding a extra breakpoint may not be a good idea.

**Comment by azro352** [comment]
> Adding a extra breakpoint may not be a good idea. Customizing the value of the existed breakpoints is our recommendation.

@afc163 so no occurences of "xxxl" at all ? Only token customization ?

**Comment by azro352** [comment]
🆘 I need help debugging the failing tests, I don't understand the real issue about the hooks call and how to fix it

- `module test (16, es, 2/2)`
```
 Summary of all failing tests
 FAIL components/_util/__tests__/responsiveObserve.test.ts
   ● Test ResponsiveObserve › test ResponsiveObserve subscribe and unsubscribe
 
     Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
     1. You might have mismatching versions of React and the renderer (such as React DOM)
     2. You might be breaking the Rules of Hooks
     3. You might have more than one copy of React in the same app
     See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
 
       62 |     theme,
       63 |     components,
     > 64 |   } = React.useContext(DesignTokenContext);
          |             ^
       65 |
       66 |   const salt = `${version}-${hashed || ''}`;
       67 |
 
       at resolveDispatcher (node_modules/react/cjs/react.development.js:1465:13)
       at Object.useContext (node_modules/react/cjs/react.development.js:1473:20)
       at useContext (components/theme/internal.tsx:64:13)
       at useResponsiveObserve (components/_util/responsiveObserve.ts:26:29)
       at Object.<anonymous> (components/_util/__tests__/responsiveObserve.test.ts:5:51)
           at runMicrotasks (<anonymous>)
```


- `test (18, dom, 2/2` (I guess it's related to the first one, and one fix only solves the both)

```
FAIL components/_util/__tests__/responsiveObserve.test.ts
  ● Test ResponsiveObserve › test ResponsiveObserve subscribe and unsubscribe

    TypeError: Cannot read properties of null (reading 'useContext')

      62 |     theme,
      63 |     components,
    > 64 |   } = React.useContext(DesignTokenContext);
         |             ^
      65 |
      66 |   const salt = `${version}-${hashed || ''}`;
      67 |

      at Object.useContext (node_modules/react/cjs/react.development.js:1618:21)
      at useContext (components/theme/internal.tsx:64:13)
      at useResponsiveObserve (components/_util/responsiveObserve.ts:26:29)
      at Object.<anonymous> (components/_util/__tests__/responsiveObserve.test.ts:5:51)
          at runMicrotasks (<anonymous>)
```

File: components/_util/__tests__/responsiveObserve.test.ts [hunk 8]
```diff
@@ -1,14 +1,15 @@
-import ResponsiveObserve, { responsiveMap } from '../responsiveObserve';
+import useResponsiveObserve from '../responsiveObserve';
 
 describe('Test ResponsiveObserve', () => {
   it('test ResponsiveObserve subscribe and unsubscribe', () => {
-    const { xs } = responsiveMap;
+    const responsiveObserve = useResponsiveObserve();
```
Comment (afc163) [inline components/_util/__tests__/responsiveObserve.test.ts:5]: You can call React hooks in here.

**Comment by azro352** [comment]
@afc163 @zombieJ @yoyo837

I think the test "ResponsiveObserve.unsubscribe should be called when unmounted" in `components/grid/__tests__/index.test.tsx`
should be deleted.
Because now we use a hook, it is not the same instance of `ResponsiveObserve` shared between the test and the Row, so we can't validate that the method have been called
do you agree ?

File: components/_util/responsiveObserve.ts [hunk 9]
```diff
@@ -1,74 +1,80 @@
...
-  xxl: '(min-width: 1600px)',
-};
+const getResponsiveMap = (token: GlobalToken): BreakpointMap => ({
+  xs: `(max-width: ${token.screenXSMax}px)`,
+  sm: `(min-width: ${token.screenSM}px)`,
+  md: `(min-width: ${token.screenMD}px)`,
+  lg: `(min-width: ${token.screenLG}px)`,
+  xl: `(min-width: ${token.screenXL}px)`,
+  xxl: `(min-width: ${token.screenXXL}px)`,
+});
```
Comment (zombieJ) [inline components/_util/responsiveObserve.ts:18]: Beautiful ❤️

**Comment by azro352** [comment]
@yoyo837 Come back 😁 the changes have been made

File: components/grid/__tests__/index.test.tsx [hunk 10]
```diff
@@ -88,15 +83,6 @@ describe('Grid', () => {
     expect(asFragment().firstChild).toMatchSnapshot();
   });
 
-  it('ResponsiveObserve.unsubscribe should be called when unmounted', () => {
```
Comment (zombieJ) [inline components/grid/__tests__/index.test.tsx:91]: Do not remove directly which should replace with new test case.

File: components/_util/responsiveObserve.ts [hunk 11]
```diff
@@ -1,74 +1,80 @@
...
 
 type SubscribeFunc = (screens: ScreenMap) => void;
 const subscribers = new Map<Number, SubscribeFunc>();
```
Comment (zombieJ) [inline components/_util/responsiveObserve.ts:21]: Could we move these `subscribers`, `subUid`, `screen` into `useResponsiveObserve`. It's no need be static variable anymore.

File: components/grid/__tests__/index.test.tsx (hunk 10 above)
Comment (azro352) [reply components/grid/__tests__/index.test.tsx:91]: I have no idea how to replace it. What do I do ? Who write the new test ?

Comment (yoyo837) [reply components/grid/__tests__/index.test.tsx:91]: Do not remove existing test cases, add some new test cases for this feature directly.

Comment (azro352) [reply components/grid/__tests__/index.test.tsx:91]: @yoyo837 @zombieJ 

I have no idea how to replace it. What do I do ? Who write the new test ?

Comment (azro352) [reply components/grid/__tests__/index.test.tsx:91]: @yoyo837 @zombieJ 
Tell me at least, the behaviour you expect to be tested ? And I will try to think about the test implementation

Comment (zombieJ) [reply components/grid/__tests__/index.test.tsx:91]: Let me handle this : )

**Review by zombieJ — APPROVED** [review]

**Review by li-jia-nan — APPROVED** [review]

**Pull request merged**